mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+using namespace mlir::arith;
+using namespace mlir::rock;
+
+namespace {
+
+// Pattern to remove redundant dtype -> newDtype -> dtype cast chains
+struct RemoveRedundantTruncExtfPattern
+    : public OpRewritePattern<linalg::GenericOp> {
+  func::FuncOp func;
+
+  RemoveRedundantTruncExtfPattern(MLIRContext *context, func::FuncOp func)
+      : OpRewritePattern<linalg::GenericOp>(context), func(func) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp generic,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Looking at generic: " << generic.getLoc() << "\n");
+
+    // Check if the generic operation only contains an ext operation.
+    if (!(isGenericWithSingleOp<arith::ExtFOp>(generic) ||
+          isGenericWithSingleOp<arith::ExtSIOp>(generic) ||
+          isGenericWithSingleOp<arith::ExtUIOp>(generic))) {
+      LLVM_DEBUG(llvm::dbgs() << "Generic doesn't contain only ext\n");
+      return failure();
+    }
+
+    bool changed = false;
+
+    auto extInput = getExtInput(generic);
+    if (!extInput)
+      return failure();
+
+    // We have checks in getExtInput to ensure that the input is not a
+    // BlockArgument, so we can assume that it has a defining op.
+    auto input = extInput.getDefiningOp();
+
+    if (isa<RockGemmWrapperInterface>(input) ||
+        isa<RockGemmGemmWrapperInterface>(input)) {
+      // Now we need to check the input of the cast (linalg.generic with ext)
+      // to see if it is a RockOp that makes use of MFMA. MFMA in rocmlir does
+      // accumulation in higher precision, so if our rock op is using MFMA and
+      // the return type isn't F32 or I32 (highest level of precision), then it
+      // means that there will be a trunc op inserted in RockGemmToGridwise
+      // that will potentially be redundant.
+      changed |= handleRockGemmWrapper(input, generic, rewriter);
+    } else if (auto inputGeneric = dyn_cast<linalg::GenericOp>(input)) {
+      // We also want to handle an arbitrary trunc op followed by ext ops
+      if (isGenericWithSingleOp<arith::TruncFOp>(inputGeneric) ||
+          isGenericWithSingleOp<arith::TruncIOp>(inputGeneric)) {
+        changed |= handleLinalgGeneric(inputGeneric, generic, rewriter);
+      }
+    } else {
+      // We have come across a pattern that we do not know how to handle, or
+      // does not require changing.
+      return failure();
+    }
+
+    return changed ? success() : failure();
+  }
+
+private:
+  // This is a helper function that checks if the generic operation contains
+  // only a single operation of type OpType, and that the yield returns that
+  // value.
+  // NOTE: A convert in migraphx, will eventually be lowered to a linalg
+  // with either a single trunc or ext op. e.g.,:
+  // %1 = migraphx.convert %0 : <1x5x3xf16, 15x3x1> to <1x5x3xf32, 15x3x1>
+  template <typename OpType>
+  bool isGenericWithSingleOp(linalg::GenericOp generic) const {
```

Comment:
We are running `linalg-elementwise-op-fusion` in highlevel pipeline which would fuse multiple linalg generic ops into a single one. 

Therefore, I don't' think it is always  guranteed that linalg will only contain single operation.  

I think "extf/truncf' are a special case because they require tensors/memrefs of different types on inputs and outputs. therefore it is probably true that linalg generic will only contains extf/truncf and it won't be fused with other linalg ops when we run `linalg-elementwise-fusion`  Do you think that's the case ? 

You can probably write a small test and run it through `linalg-elementwise-fusion`